packages/serialize/types/index.d.ts

```diff
@@ -0,0 +1,61 @@
+// Definitions by: Junyoung Clare Jang <https://github.com/Ailrun>
+// TypeScript Version: 2.8
+
+import { RegisteredCache, SerializedStyles } from '@emotion/utils';
+import * as CSS from 'csstype';
+
+export {
+  RegisteredCache,
+  SerializedStyles,
+};
+
+export type CSSProperties = CSS.PropertiesFallback<number | string>;
+export type CSSPropertiesWithMutliValues = {
+  [K in keyof CSSProperties]: CSSProperties[K] | Array<Extract<CSSProperties[K], string>>;
+};
+/**
+ * @desc Following type exists for autocompletion of key.
+ */
+export type CSSPseudos<MP> = { [K in CSS.Pseudos]?: ObjectInterpolation<MP> };
+export interface CSSOthersObject<MP> {
+  [propertiesName: string]: Interpolation<MP>;
+}
+
+export interface ComponentSelector {
+  __emotion_styles: any;
+}
+
+/**
+ * @desc Following type includes `& string` because of `animation` and `animationName`.
+ * It's hard to add exceptions only for those CSS property to make them accept `Keyframes`.
+ */
```

Comment:
What would be the point of adding exceptions when using this on anything other than animation and animationName is useless?

Reply:
@mitchellhamilton So what you said is we should add `& string` type even if it is not actual string?

Reply:
Yes, users should see it as a string.

Also, slightly unrelated but does TS have [some kind of opaque type like flow](https://flow.org/en/docs/types/opaque-types/)?

Reply:
@mitchellhamilton TypeScript (almost) always follows structural typing, so there is no equivalent. In terms of PL theory, flow is much more beautiful... The only reason I don't use it is they don't support dependencies between flow-typed typing packages... (I don't know whether now they support or not, but at least this May, they didn't support it)